src/lib/spinel/radio_spinel.cpp

```diff
@@ -785,7 +785,7 @@ void RadioSpinel::ProcessRadioStateMachine(void)
 {
     if (mState == kStateTransmitDone)
     {
-        mState        = kStateReceive;
+        mState        = mRxOnWhenIdle ? kStateReceive : kStateSleep;
```

Comment:
Hi Abtin, thank you for your quick comment!

We're currently facing some difficulties using a Host + RCP set up to join a network as an SED. After SED sends Child ID Request (and received proper ACK), RX will be off. 

More generally, if an SED sends a packet and received proper ACK, RX will be off. At this point, if mState = kStateReceive, then any calls to RadioSpinel::Receive() will not turn RX back on.

Reply:
Hi @abtink , we're looking at 2 possible scenarios (for `rx_off_when_idle` device) after receiving an ACK frame with pending bit = true:

1. Radio stays off and waits for `RadioSpinel` to send `RadioSpinel::Receive()` to receive next pending frame
    - In this case, `mState` needs to be in `kStateSleep` so that `RadioSpinel::Receive()` will turn RX back on using `Set(SPINEL_PROP_MAC_RAW_STREAM_ENABLED, SPINEL_DATATYPE_BOOL_S, true);`. `RadioSpinel::Receive()` should be called repeatedly until we have received a frame from the source of the pending bit = true. 
    - This will require `RadioSpinel` to accurately reflect Radio state (i.e. whether it is `kStateReceive` or `kStateSleep`). 
2. Radio stays on and waits for pending frame, then turns off upon receiving the next incoming frame
    - In this case, `RadioSpinel` needs to keep track of whether the next frame received is from the correct source. If it is, then no action is required (i.e. Radio will turn off RX after receiving). If the frame is not from the correct source, then need to turn RX back on again (until some timeout is reached).
    - This keeps mState = kStateReceive for `RadioSpinel`, but also requires `RadioSpinel` to be aware of the fact that Radio has gone into sleep state so that it can turn RX back on again (presumably using `RadioSpinel::Receive()` ). 

Could we seek your opinion on which method would be preferred? It seems like it may be simpler to assume that Radio will always go to sleep so that whenever a frame needs to be received, `RadioSpinel::Receive()` is called and this will always turn RX on.

Reply:
@tanyanquan ,  my first suggestion would be to avoid using `OT_RADIO_CAPS_RX_ON_WHEN_IDLE` if possible. This will be much simpler.

While PR #9554 added support for this new capability (where the radio driver itself manages sleep), it did not fully add support for this in `RadioSpinel`.

The challenge is that the `RadioSpinel` module tracks the radio state, assuming it mirrors the RCP's radio state. Without `CAPS_RX_ON_WHEN_IDLE`, the radio state transitions are simple, and the current `RadioSpinel` implementation tracks this. However, `CAPS_RX_ON_WHEN_IDLE` makes the radio state transitions more complex. To properly support this in `RadioSpinel`, I expect larger changes would be needed (effectively implementing similar logic in `RadioSpinel` as in the radio platform to handle/track all new possible state transitions).

---

Regarding your suggestion to set `mState = mRxOnWhenIdle ? kStateReceive : kStateSleep;` in `RadioSpinel` after TX is done, there are two aspects to consider:

-   It will indeed help ensure `RadioSpinel::Receive()` works correctly (since `Receive()` only sends the Spinel command to RCP if it thinks the RCP radio is asleep, i.e., `mState` is `kStateSleep`).
-   But it can cause issues if RCP receives a frame and passes it back to the host, which comes as `RadioSpinel::HandleValueIs()` for the `SPINEL_PROP_STREAM_RAW` property and calls `RadioSpinel::RadioReceive()`.
    -   In `RadioReceive()`, the `mState` is checked, and the frame would not be processed if `mState` is `kStateSleep` (on `RadioSpinel`).
    -   We can consider relaxing the check in `RadioSpinel::RadioReceive()`, but I am not sure if it may have other side effects.
- There may be other yet unknown issues if `CAPS_RX_ON_WHEN_IDLE` is used with `RadioSpinel`.